ADAL/src/ui/ADWebAuthController.m

```diff
@@ -63,538 +64,112 @@
-        }
-        else
-        {
-            MSID_LOG_ERROR(_requestParams, @"unable to redirect to browser from extension");
-        }
-#else // !TARGET_OS_IPHONE
-        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:requestURL]];
-#endif // TARGET_OS_IPHONE
-        return NO;
-    }
-    
-    // Stop at the end URL.
-    if ([[requestURL lowercaseString] hasPrefix:[_endURL lowercaseString]] ||
-        [[[request.URL scheme] lowercaseString] isEqualToString:@"msauth"])
-    {
-        // iOS generates a 102, Frame load interrupted error from stopLoading, so we set a flag
-        // here to note that it was this code that halted the frame load in order that we can ignore
-        // the error when we are notified later.
-        _complete = YES;
-        
-#if AD_BROKER
-        // If we're in the broker and we get a url with msauth that means we got an auth code back from the
-        // client cert auth flow
-        if ([[[request.URL scheme] lowercaseString] isEqualToString:@"msauth"])
-        {
-            [self webAuthDidCompleteWithURL:request.URL];
-            return NO;
-        }
+    webviewConfig.parentController = context.parentController;
 #endif
 
-        NSURL* url = request.URL;
-        [self webAuthDidCompleteWithURL:url];
-        
-        // Tell the web view that this URL should not be loaded.
-        return NO;
-    }
-    
-    // check for pkeyauth challenge.
-    if ([requestURL hasPrefix:kADALPKeyAuthUrn])
-    {
-        // We still continue onwards from a pkeyauth challenge after it's handled, so the web auth flow
-        // is not complete yet.
-        [self handlePKeyAuthChallenge:requestURL];
-        return NO;
-    }
-
-    // redirecting to non-https url is not allowed
-    if (![[[request.URL scheme] lowercaseString] isEqualToString:@"https"])
-    {
-        MSID_LOG_ERROR(nil, @"Server is redirecting to a non-https url");
-        _complete = YES;
-        ADAuthenticationError* error = [ADAuthenticationError errorFromNonHttpsRedirect:_requestParams.correlationId];
-        dispatch_async( dispatch_get_main_queue(), ^{[self endWebAuthenticationWithError:error orURL:nil];} );
-        
-        return NO;
-    }
-    
-    if ([request isKindOfClass:[NSMutableURLRequest class]])
-    {
-        [ADURLProtocol addContext:_requestParams toRequest:(NSMutableURLRequest*)request];
-    }
-    
-    return YES;
+    [MSIDWebviewAuthorization startEmbeddedWebviewAuthWithConfiguration:webviewConfig
+                                                          oauth2Factory:context.oauthFactory
+                                                                webview:context.webView
+                                                                context:requestParams
+                                                      completionHandler:completionHandler];
 }
 
-// The user cancelled authentication
-- (void)webAuthDidCancel
++ (NSDictionary *)dictFromQueryString:(NSString *)query
```

Comment:
nit: if this is still not available in common core, let's move this utility there

Reply:
This is a good point, common core's utility needs revision I think.
Query params can have key-only value and our utility does not account for this.